pkg/stub/stub.go

```diff
@@ -449,14 +457,11 @@ func (stub *stub) Run(ctx context.Context) error {
 	return err
 }
 
-// Wait for the plugin to stop.
+// Wait for the plugin to stop, should be called after Start() or Run().
 func (stub *stub) Wait() {
-	stub.Lock()
-	if stub.srvErrC == nil {
-		return
+	if stub.started {
```

Comment:
pls provide a getstarted state api... stub.Lock() started:=stub.started.. sub.Unlock() return started... at least for use when not locked..

Reply:
> pls provide a getstarted state api... stub.Lock() started:=stub.started.. sub.Unlock() return started... at least for use when not locked..

If i haven‘t misunderstood, the `getstarted` api is expect to replace all `if stub.started` code? currently many started check exist in context that have hold the lock, if provide a `getstarted` state api now may make deadlock   more likely to occur, maybe we need more consideration about this, and solve it in a new pr?
Or if your concern just the `if stub.started` without lock in `Wait()`, check `stub.started` is just a protection for not use `Wait()` before the first `Start()` or `Run()`, or else `Wait()` may try to read nil `stub.doneC`: 

- if call `Wait()` before the first `Start()` and `Run()`: the `stub.started` is must `false` and `Wait()` will not wait for the nil `stub.doneC`
- if call `Wait()` after the first `Start()` or `Run()`: there is no chance `stub.doneC` will change back to `nil`, so wait for it is safely

Reply:
For your locking vs. context of call problem you could go the conventional way of sliptting it into two functions: 

```
// IsStarted returns true if the plugin has been started either by Start() or by Run().
func (s *Stub) IsStarted() bool {
    s.Lock()
    defer s.Unlock()
    return s.isStarted()
}

func (s *Stub) isStarted() bool {
    return s.started()
}
```

Then you can simply replace your current internal `if s.started` checks in an already locked context with corresponding `if s.isStarted()` checks.

Reply:
this ^

Reply:
OK and Done.